integration/test_tor.py

```diff
@@ -125,3 +138,19 @@ def _create_anonymous_node(reactor, name, control_port, request, temp_dir, flog_
     result = yield util._run_node(reactor, node_dir.path, request, None)
     print("okay, launched")
     return result
+
+
+@pytest_twisted.inlineCallbacks
+def test_anonymous_client(reactor, alice, request, temp_dir, flog_gatherer, tor_network, introducer_furl):
+    """
+    A normal node (alice) and a normal introducer are configured, and one node
+    (anonymoose) which is configured to be anonymous by talking via Tor.
+
+    Anonymoose should be able to communicate with alice.
+
+    TODO how to ensure that anonymoose is actually using Tor?
```

Comment:
I don't know a simple way to do this.

Reply:
Since I wrote that I added some of the safety assertions in `http_client.py`'s `from_nurl()` to make it more likely to catch this.

One approach would be something like "if this codepath is hit, send a signal that can be checked". I guess Foolscap's logging something kinda has this with unique logging IDs? And you can do the same with eliot log messages, or even normal log messages by looking at strings. But that doesn't tell you whether it's maybe violating it in other ways.

Another approach is... making it impossible? Basically have an approved list of TCP and DNS connections and anything else causes a failure. This seems much more robust in that it enforces the policy, rather than trying to validate it post-hoc. I'll open a ticket.

Reply:
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4031